File: leetcode/hard/010_regular_expression_matching.cpp
# Problem: https://leetcode.com/problems/regular-expression-matching/
# Approach: 
# Time Complexity: O(M * N)
# Space Complexity: O(N)

class Solution {
public:
    bool isMatch(string s, string p) {
        int m = s.length(), n = p.length();
        vector<bool> prev(n + 1, false), curr(n + 1, false);

        prev[0] = true;
        
        for (int j = 1; j <= n; ++j) {
            if (p[j - 1] == '*') {
                prev[j] = prev[j - 2];
            }
        }

        for (int i = 1; i <= m; ++i) {
            curr[0] = false; 
            for (int j = 1; j <= n; ++j) {
                if (p[j - 1] == '.' || p[j - 1] == s[i - 1]) {
                    curr[j] = prev[j - 1];
                } else if (p[j - 1] == '*') {
                    curr[j] = curr[j - 2] || (prev[j] && (p[j - 2] == '.' || p[j - 2] == s[i - 1]));
                } else {
                    curr[j] = false;
                }
            }
            prev = curr; 
        }

        return prev[n];
    }
};

File: leetcode/hard/010_regular_expression_matching.py
# Problem: https://leetcode.com/problems/regular-expression-matching/
# Time Complexity: O(M * N)
# Space Complexity: O(N)

class Solution:
    def isMatch(self, s: str, p: str) -> bool:
        """
        :type s: str
        :type p: str
        :rtype: bool
        """
        m, n = len(s), len(p)
        prev = [False] * (n + 1)
        curr = [False] * (n + 1)

        prev[0] = True
        
        for j in range(1, n + 1):
            if p[j - 1] == '*':
                prev[j] = prev[j - 2]

        for i in range(1, m + 1):
            curr[0] = False 
            for j in range(1, n + 1):
                if p[j - 1] == '.' or p[j - 1] == s[i - 1]:
                    curr[j] = prev[j - 1]
                elif p[j - 1] == '*':
                    curr[j] = curr[j - 2] or (prev[j] and (p[j - 2] == '.' or p[j - 2] == s[i - 1]))
                else:
                    curr[j] = False
                    
            prev, curr = curr, prev 

        return prev[n]
